Fixes #15095 - introduced support for puppet environment classes api

dmitri-d opened the pull request:
- Ruby 1.8.7 is NOT supported.
  - This will attempt to cache puppet classes and rely on etag for cache validation. Enable puppet's environment class cache for this feature to have effect.
  - It will return a timeout error to the client if puppet doesn't respond within 15 seconds. The module will wait for 5 minutes to update the classes cache before timing out however.

dmitri-d committed:
Fixes #15095 - Support for puppet environment classes api

        - Ruby 1.8.7 is NOT supported.
        - This will attempt to cache puppet classes and rely on etag for cache
          validation. Enable puppet's environment class cache for this feature
          to have effect.
        - It will return a timeout error to the client if puppet doesn't
          respond within 15 seconds. The module will wait for 5 minutes to
          update the classes cache before timing out however.
        - Extended puppet api to return extended puppet classes information,
          including errors

dmitri-d commented:
Added a new api endpoint `/puppet/environments/.../classes_and_errors` that returns extended puppet classes information, including parsing errors. The new api resolves http://projects.theforeman.org/issues/2289 and should help with resolution of  http://projects.theforeman.org/issues/3900.

domcleal commented on bundler.d/puppet.rb:
```diff
@@ -3,3 +3,7 @@
 end
+
+group :puppet_proxy_puppet_api do
+  gem 'concurrent-ruby' if RUBY_VERSION > '1.8.7'
```
I'd suggest dropping 1.8.7 support entirely here and enable this always; send a PR to foreman-infra to update the test_proxy_develop.yaml file/job and remove it from there.

domcleal commented on modules/puppet_proxy/puppet_api.rb:
```diff
@@ -45,8 +45,26 @@ class Proxy::Puppet::Api < ::Sinatra::Base
+    rescue Proxy::Puppet::TimeoutError => e
+      log_halt 503, e.message
+    rescue Exception => e
+      p "!!! #{e.class}"
```
Pass `e` to `log_halt` instead (and anywhere the full exception will be useful).

domcleal commented on modules/puppet_proxy_common/api_request.rb:
```diff
@@ -13,11 +13,12 @@ def initialize(url, ssl_ca, ssl_cert, ssl_key)
       http             = Net::HTTP.new(uri.host, uri.port)
       http.use_ssl     = uri.scheme == 'https'
       http.verify_mode = OpenSSL::SSL::VERIFY_NONE
+      http.read_timeout = timeout
```
Nitpick, alignment is out.

domcleal commented on modules/puppet_proxy_puppet_api/plugin_configuration.rb:
```diff
@@ -16,15 +16,35 @@ def load_classes
       require 'puppet_proxy_puppet_api/v3_api_request'
       require 'puppet_proxy_puppet_api/v3_environments_retriever'
       require 'puppet_proxy_puppet_api/v3_classes_retriever'
+      require 'puppet_proxy_puppet_api/v3_environmentclassesapi_classes_retriever' if RUBY_VERSION > '1.8.7'
```
ditto, drop 1.8.7 and load this always

domcleal commented on modules/puppet_proxy_puppet_api/plugin_configuration.rb:
```diff
@@ -16,15 +16,35 @@ def load_classes
+        container_instance.dependency :class_cache_initializer, lambda {Proxy::PuppetApi::NoopClassesCacheInitializer.new}
+      end
 
+      if settings[:puppet_version].to_s >= "4.4" && RUBY_VERSION > '1.8.7'
```
Prefer `else`, as this is just an inversion of the previous conditional I think.

domcleal commented on modules/puppet_proxy_puppet_api/v3_environmentclassesapi_classes_retriever.rb:
```diff
@@ -0,0 +1,100 @@
+      begin
+        response, etag = @puppet_api.new(puppet_url, ssl_ca, ssl_cert, ssl_key).list_classes(environment, etag, MAX_PUPPETAPI_TIMEOUT)
+      rescue Exception => e
+        logger.error("Error while retrieving puppet classes for '%s' environment: %s" % [environment, e])
```
It might be best to pass e.backtrace to logger.error here, for completeness.

domcleal commented on modules/puppet_proxy_puppet_api/v3_environmentclassesapi_classes_retriever.rb:
```diff
@@ -0,0 +1,100 @@
+require 'concurrent'
+
+class Proxy::PuppetApi::V3EnvironmentClassesApiClassesRetriever
```
The filename is missing some underscores to match the capitalisation of the class, it'd be easier to read with them.

domcleal commented on modules/puppet_proxy_puppet_api/v3_environmentclassesapi_classes_retriever.rb:
```diff
@@ -0,0 +1,100 @@
+  end
+
+  def start_service
+    @environments.all.map(&:name).each {|environment| @classes.get_classes(environment)}
```
Could you add some logging here so it's clear why the requests are being made at startup? Perhaps also a log message to state when it's complete.

domcleal replied:
Secondly, if the smart proxy receives an API request shortly after startup while this preloading is occurring, it looks like there will be multiple requests sent to the Puppet server. Can this be limited so the request is blocked until the previous query has resolved or has timed out?

dmitri-d commented:
- Dropped conditional loading of dependencies (assumes we are on ruby > 1.8.7), and fixed the rest of the issues pointed out in the feedback.
- Concurrent calls to retrieve classes from the same environment result in only one actual call to puppet.

domcleal commented on bundler.d/puppet.rb:
```diff
@@ -3,3 +3,7 @@
 end
+
+group :puppet_proxy_puppet_api do
+  gem 'concurrent-ruby'
```
Pin to `~> 1.0`, it claims to obey semver.

domcleal commented on modules/puppet_proxy_puppet_api/v3_environment_classes_api_classes_retriever.rb:
```diff
@@ -0,0 +1,115 @@
+      begin
+        response, etag = @puppet_api.new(puppet_url, ssl_ca, ssl_cert, ssl_key).list_classes(environment, etag, MAX_PUPPETAPI_TIMEOUT)
+      rescue Exception => e
+        logger.error("Error while retrieving puppet classes for '%s' environment: %s" % [environment, e.backtrace])
```
This isn't logging the actual error, only the backtrace and environment name. Use `logger.error("Error while retrieving puppet classes for '%s' environment" % environment, e)` to log the full exception info and with better formatting.

domcleal commented on modules/puppet_proxy_puppet_api/v3_environment_classes_api_classes_retriever.rb:
```diff
@@ -0,0 +1,115 @@
+
+  def start
+    logger.info("Started puppet class cache initialization")
+    futures = @environments.all.map(&:name).map do |environment|
```
The environment retrieval here is in the main service startup thread, so when the Puppet Server's inaccessible or slow, the module fails to load. Can you move it to a thread?

<pre>
E, [2016-09-23T11:58:55.600886 #2886] ERROR -- : Couldn't enable 'puppet_proxy_puppet_api': Failed to open TCP connection to foreman.example.com:8140 (Connection refused - connect(2) for "foreman.example.com" port 8140)
D, [2016-09-23T11:58:55.600921 #2886] DEBUG -- : ["/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/net/http.rb:882:in `rescue in block in connect'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/net/http.rb:879:in `block in connect'", "/home/dcleal/.rvm/rubies/ru
by-2.3.0/lib/ruby/2.3.0/timeout.rb:91:in `block in timeout'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/timeout.rb:101:in `timeout'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/net/http.rb:878:in `connect'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby
/2.3.0/net/http.rb:863:in `do_start'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/net/http.rb:852:in `start'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/net/http.rb:1398:in `request'", "/home/dcleal/code/foreman/smart-proxy/modules/puppet_proxy_common/a
pi_request.rb:39:in `send_request'", "/home/dcleal/code/foreman/smart-proxy/modules/puppet_proxy_puppet_api/v3_api_request.rb:4:in `find_environments'", "/home/dcleal/code/foreman/smart-proxy/modules/puppet_proxy_common/environments_retriever_base.rb:18:in `all'", "/home/
dcleal/code/foreman/smart-proxy/modules/puppet_proxy_puppet_api/v3_environment_classes_api_classes_retriever.rb:109:in `start'", "/home/dcleal/code/foreman/smart-proxy/lib/proxy/plugin_initializer.rb:228:in `block in start_services'", "/home/dcleal/code/foreman/smart-prox
y/lib/proxy/plugin_initializer.rb:226:in `each'", "/home/dcleal/code/foreman/smart-proxy/lib/proxy/plugin_initializer.rb:226:in `start_services'", "/home/dcleal/code/foreman/smart-proxy/lib/proxy/plugin_initializer.rb:211:in `configure_plugin'", "/home/dcleal/code/foreman
</pre>


(This is very likely at server startup as PS is much slower than foreman-proxy to load.)

dmitri-d commented:
Responded to the feedback. Cache initialization is completely async now. Cache initialization failures are being logged now.

domcleal commented on modules/puppet_proxy_puppet_api/v3_environment_classes_api_classes_retriever.rb:
```diff
@@ -0,0 +1,121 @@
+        response, etag = @puppet_api.new(puppet_url, ssl_ca, ssl_cert, ssl_key).list_classes(environment, etag, MAX_PUPPETAPI_TIMEOUT)
+      rescue Exception => e
+        logger.error("Error while retrieving puppet classes for '%s' environment" % [environment], e)
+        raise e
```
If an error is raised here, it looks like `@futures_cache[environment]` doesn't get reset to nil, resulting in permanent failures from the smart proxy. I had a transient error, which was logged with a stack trace on line 78, but subsequent calls to the API immediately return the same error message (execution expired).

dmitri-d replied:
Nice catch indeed.

dmitri-d commented:
Futures cache is now being cleared if an exception occurs when talking to puppet server. Added a test to verify cache clearing on error.

domcleal commented:
There's a rubocop failure, can you fix it please?

dmitri-d commented:
Unhappy rubocop should be happy now.

domcleal commented:
Merged as 8799d0c, thanks @witlessbird.

stbenjam commented:
On el7 nightly, getting this in the proxy.log:

```
E, [2016-09-30T16:15:22.213297 #2456] ERROR -- : Disabling all modules in the group ['puppet_proxy_puppet_api', 'puppet'] due to a failure in one of them: Gem loading error: cannot load such file -- concurrent/ruby
```

Filed http://projects.theforeman.org/issues/16760
